Guides/Grouped.md

```diff
@@ -0,0 +1,68 @@
+# AdjacentPairs
+
+[[Source](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Grouped.swift) | 
+ [Tests](https://github.com/apple/swift-algorithms/blob/main/Tests/SwiftAlgorithmsTests/GroupedTests.swift)]
+ 
+Groups up elements of a sequence into a new Dictionary, whose values are Arrays of grouped elements, each keyed by the result of the given closure.
+
+```swift
+let fruits = ["Apricot", "Banana", "Apple", "Cherry", "Avocado", "Coconut"]
+let fruitsByLetter = fruits.grouped(by: { $0.first! })
+// Results in:
+// [
+//     "B": ["Banana"],
+//     "A": ["Apricot", "Apple", "Avocado"],
+//     "C": ["Cherry", "Coconut"],
+// ]
+```
+
+If you wish to achieve a similar effect but for single values (instead of Arrays of grouped values), see [`keyed(by:)`](Keyed.md).
+
+## Detailed Design
+
+The `grouped(by:)` method is declared as a `Sequence` extension returning
+`[GroupKey: [Element]]`.
+
+```swift
+extension Sequence {
+    public func grouped<GroupKey>(
+        by keyForValue: (Element) throws -> GroupKey
+    ) rethrows -> [GroupKey: [Element]]
+}
+```
+
+### Complexity
+
+Calling `grouped(by:)` is an O(_n_) operation.
+
+### Comparison with other languages
+
+| Language      | Grouping API |
+|---------------|--------------|
+| Java          | [`groupingBy`](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/util/stream/Collectors.html#groupingBy(java.util.function.Function)) |
+| Kotlin        | [`groupBy`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/group-by.html) |
+| C#            | [`GroupBy`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=net-7.0#system-linq-enumerable-groupby) |
+| Rust          | [`group_by`](https://doc.rust-lang.org/std/primitive.slice.html#method.group_by) |
+| Ruby          | [`group_by`](https://ruby-doc.org/3.2.2/Enumerable.html#method-i-group_by) |
+| Python        | [`groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby) |
+| PHP (Laravel) | [`groupBy`](https://laravel.com/docs/10.x/collections#method-groupby) |
+
+#### Naming
+
+All the surveyed languages name this operation with a variant of "grouped" or "grouping". The past tense `grouped(by:)` best fits [Swift's API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/).
+
+#### Customization points
+
+Java and C# are interesting in that they provide multiple overloads with several points of customization:
+
+1. Changing the type of the groups.
+    1. E.g. the groups can be Sets instead of Arrays.
+    1. Akin to calling `.transformValues { group in Set(group) }` on the resultant dictionary, but avoiding the intermediate allocation of Arrays of each group.
```

Comment:
### Forward

I explored all this really just for my own edification, but it might be helpful to include some summary of this in your notes here, to explain _why_ the `group(by:)` API doesn't and _cannot_ currently support this functionality.  Adding all the necessary infrastructure - whether a `Collector` protocol or whatever else - seems like a more fundamental addition to the Swift standard library that should be tackled in its own patch & pitch.

### Rabbit hole

I like the idea of supporting this, but it doesn't seem like it's possible [elegantly] in Swift today…?  Swift's standard library & Foundation don't include an equivalent to Java's `Collector` interface, for example.  And even though I'm not a big fan of that approach anyway, you fundamentally do need a way to say "any collection that can be initialised and added to", which Swift does not have.

Specifically: you can easily genericise `group(by:)` with some `ValueCollection` type, but what protocol would it have to conform to?  `Collection` specifies only the APIs for _reading_ from a _given_ collection, not for constructing one or adding to it.  `MutableCollection` (surprisingly) has the exact same limitations, _and_ feels inappropriate anyway since it forces the resulting Dictionary to have a mutable type for its values.

There is `RangeReplaceableCollection` as a half measure - it does include an init method, but only a limited subset of standard collections conform to it (notably not including `Set`).  Still, with that you can do:

```swift
extension Sequence {
    public func grouped<GroupKey, ValueCollection: RangeReplaceableCollection<Element>>(
        by keyForValue: (Element) throws -> GroupKey
    ) rethrows -> [GroupKey: ValueCollection] {
        var result = [GroupKey: ValueCollection]()

        for value in self {
            result[try keyForValue(value), default: ValueCollection()].append(value)
        }

        return result
    }
}
```

Semantically it's a hack, of course - there's nothing about the _range replacement_ functionality actually needed here, it's just used because it exposes other APIs almost as a side-effect.

And anyway, the ergonomics are poor since Swift doesn't support specifying a default value for a generic parameter, requiring boilerplate in many cases:

```swift
[1, 2, 3, 4].grouped { $0 % 2 } // ❌ Generic parameter 'ValueCollection' could not be inferred

let result: [Int: [Int]] = [1, 2, 3, 4].grouped { $0 % 2 } // Works, but awkward for many use-cases.
```

One could work around the initialisation problem by having the caller explicitly provide an initialiser, but you still need a protocol that all collections support which specifies some kind of `append` method, and that doesn't currently exist.

It _is_ possible to support arbitrary collection types, but only by having the caller provide a reducer explicitly in order to work around the aforementioned limitations, e.g.:

```swift
extension Sequence {
    public func grouped<GroupKey, ValueCollection: Collection>(
        by keyForValue: (Element) throws -> GroupKey,
        collectionInitialiser: () -> ValueCollection = Array<Element>.init,
        collectionReducer: (inout ValueCollection, Element) -> ValueCollection
    ) rethrows -> [GroupKey: ValueCollection] where ValueCollection.Element == Element {
        var result = [GroupKey: ValueCollection]()

        for value in self {
            collectionReducer(&result[try keyForValue(value), default: collectionInitialiser()], value)
        }

        return result
    }
}

let result: [Int: [Int]] = [1, 2, 3, 4].grouped(by: { $0 % 2 },
                                                collectionReducer: { $0.append($1); return $0 })
```

This is essentially now a functional superset of `group(by:)` and `keyed(by:)`.

You can't provide a default reducer since there's no way to generically add a value to any type of collection (per the earlier point).  So this'd have to be a special parallel version of group(by:), in addition to the 'simple' one that just hard-codes use of `Array` and doesn't require that a reducer be specified.

I'm not a fan of this approach, technically capable as it may be, though admittedly that's from a subjective standpoint.  I think it'd be more elegant, and cleaner for library authors, to evolve Swift (or the Swift standard library) to better support generalising across all collection types.

Reply:
Very interesting findings, thanks for sharing this with me! How do you think it would be best to communicate these in the library? Perhaps an addendum to the new md files I'm adding in this PR?

> but you still need a protocol that all collections support which specifies some kind of append method, and that doesn't currently exist.

Oh wow, I never noticed this, but you're right!

> It is possible to support arbitrary collection types, but only by having the caller provide a reducer explicitly in order to work around the aforementioned limitations, e.g.:

This is what I had in mind (or use a new protocol with method requirements, instead of two separate closure params).

This is more general, because the groups might not be collections at all. The histogram example comes to mind, where you'd want something like:

```
[1, 2, 3, 4].grouped(
    by: { $0 % 2 },
    collectionInitialiser: { 0 }, // not a collection, but w/e
    collectionReducer: { counter, _ in counter += 1 },
)
```

If there were a protocol for this, then there might be some `CountingCollector` that could be used.

Reply:
I'm not sure how best to summarise that - obviously what I wrote, as a kind of deep-dive on this tangent, is way too much for these documents.  Maybe it suffices to say something like "This would require either (a) a new protocol, supported by all standard library collection types, which represents an initialisable object that can be added to, or (b) a variant method which accepts a custom 'reducer' closure".  If you want to say anything about it at all.  Perhaps this thread here on the pull request is sufficient documentation for posterity (which could be hyperlinked to from the markdown file).

Having used Java's version of this a few times in real-world code, my main comment is that these sorts of APIs - groupby etc - should never _require_ explicit specification of the return types (or a custom reducer).  They can _allow_ it, through overloads or optional parameters, but it needs to work effortlessly for the vastly most common case of just returning a Dictionary of Arrays.  A lot of Java APIs explicitly require you to pass a collector which is almost always `Collectors.toList()`… it gets very repetitive, and verbose.

Reply:
Also, as a counter-argument (Devil's advocate style) for the use of this custom-reducer `grouped(by:)` variant, you can achieve the same thing with:

```swift
[1, 2, 3, 4].reduce(into: [Int: Int]()) { $0[$1, default: 0] += 1 }
```

(and the explicit typing of the `Result` type can be omitted in some cases where it's implied by surrounding context, making this even terser)

Personally I don't mind if a language or library provides multiple ways to achieve the same goals, but I'm not sure a fancier `group(by:)` adds any benefit over the existing `reduce(into:_:)` approach.

That said, you can implement `group(by:)` using `reduce(into:_:)` too, e.g.:

```swift
[1, 2, 3, 4].reduce(into: [Int: [Int]]()) {
    $0[$1 % 2, default: []].append($1)
}
```

…yet I still feel like a purpose-built `group(by:)` is worth having.  Perhaps it's technically just a "convenience", a bit more ergonomic - but it feels like it's worth it for this simple, common case.  I guess I'm just not sure where to draw the line between a more powerful `group(by:)` and just having people fall back to `reduce(into:_:)` for uncommon needs.

Counter-counter-argument:  even though `reduce(_:_:)` and `reduce(into:_:)` _can_ be used for a huge variety of tasks, I usually find them hard to interpret when reading the code.  A large number of `Sequence` methods can be implemented using just the reducer methods - e.g. `allSatisfy`, `min` / `max`, and even `map`! - yet we still have all those purpose-built methods.  And I think Swift is better for them - they make code much more readable, even if they're technically redundant.

Reply:
> I guess I'm just not sure where to draw the line between a more powerful `group(by:)` and just having people fall back to `reduce(into:_:)` for uncommon needs.

Likewise. I have conviction about the base `group(by:)` and its semantics, but I don't know how far its extensibility should go.

>Counter-counter-argument: even though `reduce(_:_:)` and `reduce(into:_:)` can be used for a huge variety of tasks, I usually find them hard to interpret when reading the code. A large number of `Sequence` methods can be implemented using just the reducer methods - e.g. `allSatisfy`, `min` / `max`, and even `map`! - yet we still have all those purpose-built methods. And I think Swift is better for them - they make code much more readable, even if they're technically redundant.

Heh I have an entire little blog post about this: https://github.com/amomchilov/Blog/blob/master/Don't%20abuse%20reduce.md

Reply:
Is this research something you'd like to commit? You deserve to have your name on your findings :)